### PR TITLE
[breaking] Change internal model serialization to UBJSON.

### PR DESCRIPTION
**Pull request by trivialfis**
* Use typed array for models.
* Change the memory snapshot format.
* Add new C API for saving to raw format.

This PR breaks the schema of JSON, the `default_left` field is changed from boolean to integer so that we can use typed array.  XGBoost itself will continue to load the old models, third-party libraries that depend on this schema will have to make changes accordingly.  I can help with treelite once the full PR is merged.

Extracted from https://github.com/dmlc/xgboost/pull/7545 .

### CODE REVIEW

**Comment by hcho3** [comment]
> I can help with treelite once the full PR is merged.

Treelite doesn't need to be updated, since it already handles integer arrays for `default_left`. See dmlc/treelite#337.

**Comment by trivialfis** [comment]
Ah, the parser is interesting, I should learn more about it.  Thanks for sharing.

**Review by hcho3 — CHANGES_REQUESTED** [review]

File: include/xgboost/c_api.h [hunk 1]
```diff
@@ -874,8 +874,8 @@ XGB_DLL int XGBoosterPredict(BoosterHandle handle,
  *
  * \param handle Booster handle
  * \param dmat   DMatrix handle
- * \param c_json_config String encoded predict configuration in JSON format, with
- *                      following available fields in the JSON object:
+ * \param json_config String encoded predict configuration in JSON format, with
+ *                    following available fields in the JSON object:
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply include/xgboost/c_api.h:878]: Let me change that.

[the remark this answers is not in the document]
Comment (trivialfis) [reply include/xgboost/c_api.h]: Reverted.

**Comment by trivialfis** [comment]
@hcho3 Could you please take another look?

**Comment by trivialfis** [comment]
Added an extra optimization to avoid the i64 array for split index.  Also, fixed the type used in the tree param.

**Review by RAMitchell — APPROVED** [review]

**Pull request merged**